metricflow/sql/optimizer/cte_alias_to_cte_node_mapping.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+
+import logging
+from typing import Dict
+
+from metricflow_semantics.mf_logging.lazy_formattable import LazyFormat
+
+from metricflow.sql.sql_plan import SqlCteAliasMapping, SqlSelectStatementNode
+
+logger = logging.getLogger(__name__)
+
+
+class SqlCteAliasMappingLookup:
+    """A mutable lookup that stores the CTE-alias mapping at a given node.
+
+    In cases with nested CTEs in a SELECT, it's possible that a CTE defined in an inner SELECT has an alias that is the
```

Comment:
Why would they be defined with the same alias? Shouldn't we use something similar to `_next_unique_table_alias()` to build CTE aliases? Or is the problem that both the `DataflowToSqlQueryPlanConverter` and the optimizer won't have access to that function to ensure unique aliases?

Reply:
Gotcha. I definitely was planning to use something similar to `_next_unique_table_alias()` in `DataflowToSqlQueryPlanConverter` for CTE aliases. If you have a strong opinion about handling this defensively, that's fine and we can leave it here. Just might be nice to simplify the logic if it's not necessary!

Reply:
My aim is to try to maintain the contract: if the optimizer is fed in valid SQL, then it should output valid SQL. Otherwise, it becomes harder to use / makes a trap.

Reply:
Ohhh ok I think I was misunderstanding. I was imagining the CTEs would always be bubbled up to the top level, even if there were defined when visiting a subquery. I didn't realize that you could have a CTE that lives inside a subquery. In that case, your logic makes sense.

Reply:
Yeah, I was thinking about your use case and thought `CTE that lives inside a subquery` was going to be a better fit since it was only locally needed? There's flexibility though, so let me merge this so that you can try it out.

Reply:
Yeah, either should be fine from an execution standpoint, but keeping the CTEs at the top level might be most readable!